src/Build/Globbing/MSBuildGlob.cs

```diff
@@ -198,11 +203,28 @@ public static MSBuildGlob Parse(string globRoot, string fileSpec)
                         return (normalizedFixedPart, wildcardDirPart, filePart);
                     });
 
-                // compile the regex since it's expected to be used multiple times
-                var regex = isLegalFileSpec
-                    ? new Regex(matchFileExpression, FileMatcher.DefaultRegexOptions | RegexOptions.Compiled)
-                    : null;
+                Regex regex = null;
+                if (isLegalFileSpec)
+                {
+                    lock (s_regexCache)
+                    {
+                        s_regexCache.TryGetValue(matchFileExpression, out regex);
+                    }
 
+                    if (regex == null)
+                    {
+                        // compile the regex since it's expected to be used multiple times
```

Comment:
As someone complete unfamiliar with the code and just a drive-by-reviewer, I am wondering why this code block inside the `if` doesn't simply create the regex and assigns it uncoditionally to the WeakValueDictionary. From what I undestood, the goal here is to share the regex and IMHO it shouldn't matter which of the diffrent regex instances are used, as long as they both match the same regex pattern and have the same options.

If my understading is correct, this much simpler version for the body of the `if (regex == null)` would also work.

```
Regex regex = new Regex(matchFileExpression, FileMatcher.DefaultRegexOptions | RegexOptions.Compiled);
lock (s_regexCache) {
    s_regexCache[matchFileExpression] = regex;
}
```

I must be missing something obvious. Care to explain?

Reply:
It is true that functionally the code you are proposing would work just as well. Performance-wise, though, if multiple threads create equivalent `Regex` instances, it is better if the first one wins and is used by all threads as opposed to overwriting the dictionary entry (i.e. last one wins) which, in the worst case, could end up with a different instance on each thread, so effectively zero sharing. With the code in this PR, if a thread creates a `Regex` and then finds that it's lost the race, the new `Regex` won't be JIT-compiled and likely collected in a gen0 GC so pretty cheap.

In summary, I believe that the cost of the extra dictionary lookup is negligible compared to the cost of newing up and fully compiling the `Regex` so worth doing here to guarantee that there is at most one `Regex` instance for the given match file expressions.

Reply:
@ladipro Thx for the explanation, it clears things up and I agree with your reasoning.